# Lab 6 - AutoML implementation using AutoGluon

## Install the AutoGluon package

In [11]:
!python -m pip install autogluon

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.2/41.2 kB 1.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 7.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.9/116.9 kB 11.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 11.6 MB/s eta 0:00:00:00:01
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.5/49.5 kB 7.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 85.1/85.1 kB 12.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 136.5/136.5 kB 21.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 117.0/117.0 kB 22.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.1

## Required import

In [13]:
from autogluon.tabular import TabularDataset, TabularPredictor
from sklearn.model_selection import train_test_split

## Load the dataset into AutoGluon's dataset object

We have a [dataset](https://www.kaggle.com/datasets/ayessa/salary-prediction-classification) that holds information about an individual such as their age, workclass, education, marital-status, occupation, relationship status, race, sex, capital-loss, job hours, and native country.

The target class is called salary. It classifies whether a person has a salary <=50K or >50K

In [45]:
train_data = TabularDataset('salary.csv')

Loaded data from: salary.csv | Columns = 15 / 15 | Rows = 32561 -> 32561


In [47]:
train_data.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  salary  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

## Split the data into a train and validation sets

In [48]:
train, test = train_test_split(train_data, test_size=0.3, shuffle=True)

In [49]:
train.head()

age   workclass  fnlwgt      education  education-num  \
32098   45     Private  170871        HS-grad              9   
25206   47   State-gov  108890        HS-grad              9   
23491   48     Private  187505   Some-college             10   
12367   29     Private  145592        HS-grad              9   
7054    23     Private  203003        7th-8th              4   

            marital-status     occupation    relationship    race      sex  \
32098   Married-civ-spouse   Craft-repair         Husband   White     Male   
25206             Divorced   Adm-clerical       Unmarried   White   Female   
23491   Married-civ-spouse          Sales         Husband   White     Male   
12367        Never-married   Craft-repair   Not-in-family   White     Male   
7054         Never-married   Craft-repair   Not-in-family   White     Male   

       capital-gain  capital-loss  hours-per-week  native-country  salary  
32098          7298             0              60   United-States    >50K  
25206          1831             0              38   United-States   <=50K  
23491             0             0              50   United-States    >50K  
12367             0             0              40       Guatemala   <=50K  
7054              0             0              25         Germany   <=50K

In [50]:
test.head()

age workclass  fnlwgt      education  education-num  \
22278   27   Private  177119   Some-college             10   
8950    27   Private  216481      Bachelors             13   
7838    25   Private  256263     Assoc-acdm             12   
16505   46   Private  147640        5th-6th              3   
19140   45   Private  172822           11th              7   

            marital-status         occupation    relationship  \
22278             Divorced       Adm-clerical       Unmarried   
8950         Never-married     Prof-specialty   Not-in-family   
7838    Married-civ-spouse              Sales         Husband   
16505   Married-civ-spouse   Transport-moving         Husband   
19140             Divorced   Transport-moving   Not-in-family   

                      race      sex  capital-gain  capital-loss  \
22278                White   Female             0             0   
8950                 White   Female             0             0   
7838                 White     Male             0             0   
16505   Amer-Indian-Eskimo     Male             0          1902   
19140                White     Male             0          2824   

       hours-per-week  native-country  salary  
22278              44   United-States   <=50K  
8950               40   United-States   <=50K  
7838               40   United-States   <=50K  
16505              40   United-States   <=50K  
19140              76   United-States    >50K

## Modeling

* We are specifying the evaluation metric to be used as **F1 score.**
* We will save the trained model into a folder called **AutoGluonModels.**
* We are specifying the name of the target variable which is **salary.**

In [53]:
# The evaluation metric used is F1 Score
EVAL_METRIC = "f1"

# Where trained models are saved
SAVE_PATH = "AutoGluonModels"

# Column to predict
LABEL = "salary"

## Initializing AutoGluon's TabularPredictor

In [54]:
predictor  = TabularPredictor(label=LABEL, path=SAVE_PATH, eval_metric=EVAL_METRIC)

[TabularPredictor](https://auto.gluon.ai/stable/api/autogluon.tabular.TabularPredictor.html) is a class that predicts values in a column of a tabular dataset. It can be used for both classification and regression problems. In this project, we are using classification.

## Fit the predictor on the training set

In [55]:
predictor = predictor.fit(train)

No presets specified! To achieve strong results with AutoGluon, it is recommended to use the available presets.
	Recommended Presets (For more details refer to https://auto.gluon.ai/stable/tutorials/tabular/tabular-essentials.html#presets):
	presets='best_quality'   : Maximize accuracy. Default time_limit=3600.
	presets='high_quality'   : Strong accuracy with fast inference speed. Default time_limit=3600.
	presets='good_quality'   : Good accuracy with very fast inference speed. Default time_limit=3600.
	presets='medium_quality' : Fast training time, ideal for initial prototyping.
Beginning AutoGluon training ...
AutoGluon will save models to "AutoGluonModels"
=================== System Info ===================
AutoGluon Version:  1.0.0
Python Version:     3.10.13
Operating System:   Linux
Platform Machine:   x86_64
Platform Version:   #1 SMP Tue Feb 13 18:46:41 UTC 2024
CPU Count:          2
Memory Avail:       2.42 GB / 3.76 GB (64.3%)
Disk Space Avail:   4.51 GB / 4.78 GB (94.4%)
	We

### Retrieve the performance of the fitted models

In [56]:
predictor.leaderboard(silent=True)

model  score_val eval_metric  pred_time_val    fit_time  \
0   WeightedEnsemble_L2   0.725681          f1       0.983975  168.599248   
1               XGBoost   0.714426          f1       0.079966    9.025889   
2              LightGBM   0.707510          f1       0.041574    2.766567   
3         LightGBMLarge   0.704390          f1       0.091530    4.075467   
4              CatBoost   0.701195          f1       0.030191   66.034918   
5            LightGBMXT   0.693069          f1       0.066486    3.741347   
6       NeuralNetFastAI   0.692511          f1       0.063648   28.114774   
7      RandomForestGini   0.689520          f1       0.177135    5.921120   
8      RandomForestEntr   0.683845          f1       0.193663    7.545258   
9        NeuralNetTorch   0.683206          f1       0.053046   41.236620   
10       ExtraTreesGini   0.667327          f1       0.229027    4.239522   
11       ExtraTreesEntr   0.666008          f1       0.423944    5.473397   
12       KNeighborsDist   0.424051          f1       0.033588    0.124770   
13       KNeighborsUnif   0.378698          f1       0.070191    0.132563   

    pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  \
0                 0.006414           4.326358            2       True   
1                 0.079966           9.025889            1       True   
2                 0.041574           2.766567            1       True   
3                 0.091530           4.075467            1       True   
4                 0.030191          66.034918            1       True   
5                 0.066486           3.741347            1       True   
6                 0.063648          28.114774            1       True   
7                 0.177135           5.921120            1       True   
8                 0.193663           7.545258            1       True   
9                 0.053046          41.236620            1       True   
10                0.229027           4.239522            1       True   
11                0.423944           5.473397            1       True   
12                0.033588           0.124770            1       True   
13                0.070191           0.132563            1       True   

    fit_order  
0          14  
1          11  
2           4  
3          13  
4           7  
5           3  
6          10  
7           5  
8           6  
9          12  
10          8  
11          9  
12          2  
13          1

Notes: **silent=True** suppresses print ouput when we run the cell

We can see from the results above a list of models that were fitted on the training set. As specified above, the evaluation metric used was F1 Score, and it appears that the model with the best F1 score is **WeightedEnsembre_L2**. It has the highest score which is 0.725681, followed by XGBoost with an F1 score of 0.714426.

## Evaluation

**extra_info** provides information on hyperparameters and the features for fitting.

In [57]:
test_leaderboard = predictor.leaderboard(test, silent=True, extra_info=True)

In [58]:
test_leaderboard

model  score_test  score_val eval_metric  pred_time_test  \
0   WeightedEnsemble_L2    0.718351   0.725681          f1        2.517877   
1               XGBoost    0.716505   0.714426          f1        0.284374   
2              CatBoost    0.715429   0.701195          f1        0.089407   
3              LightGBM    0.711232   0.707510          f1        0.147565   
4         LightGBMLarge    0.706709   0.704390          f1        0.276871   
5            LightGBMXT    0.701381   0.693069          f1        0.265850   
6       NeuralNetFastAI    0.700163   0.692511          f1        0.213749   
7        NeuralNetTorch    0.684431   0.683206          f1        0.069568   
8      RandomForestGini    0.680092   0.689520          f1        0.565587   
9      RandomForestEntr    0.678169   0.683845          f1        0.572104   
10       ExtraTreesEntr    0.665286   0.666008          f1        0.859635   
11       ExtraTreesGini    0.662833   0.667327          f1        0.872819   
12       KNeighborsDist    0.441255   0.424051          f1        0.067879   
13       KNeighborsUnif    0.395336   0.378698          f1        0.089007   

    pred_time_val    fit_time  pred_time_test_marginal  \
0        0.983975  168.599248                 0.004604   
1        0.079966    9.025889                 0.284374   
2        0.030191   66.034918                 0.089407   
3        0.041574    2.766567                 0.147565   
4        0.091530    4.075467                 0.276871   
5        0.066486    3.741347                 0.265850   
6        0.063648   28.114774                 0.213749   
7        0.053046   41.236620                 0.069568   
8        0.177135    5.921120                 0.565587   
9        0.193663    7.545258                 0.572104   
10       0.423944    5.473397                 0.859635   
11       0.229027    4.239522                 0.872819   
12       0.033588    0.124770                 0.067879   
13       0.070191    0.132563                 0.089007   

    pred_time_val_marginal  fit_time_marginal  ...  \
0                 0.006414           4.326358  ...   
1                 0.079966           9.025889  ...   
2                 0.030191          66.034918  ...   
3                 0.041574           2.766567  ...   
4                 0.091530           4.075467  ...   
5                 0.066486           3.741347  ...   
6                 0.063648          28.114774  ...   
7                 0.053046          41.236620  ...   
8                 0.177135           5.921120  ...   
9                 0.193663           7.545258  ...   
10                0.423944           5.473397  ...   
11                0.229027           4.239522  ...   
12                0.033588           0.124770  ...   
13                0.070191           0.132563  ...   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    hyperparameters  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                        {'use_orig_features': False, 'max_base_models': 25, 'max_base_models_per_type': 5, 'save_bag_folds': True}   
1                                                              

## Making predictions

In [63]:
y_pred = predictor.predict(test.drop(LABEL, axis=1))
y_pred.head()  # Predictions

22278     <=50K
8950      <=50K
7838      <=50K
16505     <=50K
19140      >50K
Name: salary, dtype: object

The results above shows a sample of predictions of 5 observations which have the displayed row index.

In [64]:
y_pred_proba = predictor.predict_proba(test.drop(LABEL, axis=1))
y_pred_proba.head()

<=50K      >50K
22278  0.988712  0.011288
8950   0.978620  0.021380
7838   0.797905  0.202095
16505  0.686469  0.313531
19140  0.472173  0.527827

`predict_proba` returns a sample with the probabilities. 
For example, the trained model predicts that the observation with index 22278 has around 98% chance of having a salary<50k and around 1% chance of having a salary>50K. It took the highest percentage (98%) which is salary<=50K.

### Returning all the accuracy scores from the predictions

Evaluating the predictor on the lalebed test data

In [65]:
perf = predictor.evaluate_predictions(y_true=test[LABEL], y_pred=y_pred_proba)

In [66]:
perf

{'f1': 0.7183508850548959,
 'accuracy': 0.8713276691575391,
 'balanced_accuracy': 0.8057142264331736,
 'mcc': 0.637166135283908,
 'roc_auc': 0.9273684629284584,
 'precision': 0.7629700142789148,
 'recall': 0.678662150719729}

We have:
* An F1 Score of around 0.71 / 71%
* Around 87% accuracy
* An ROC_AUC of about 0.92
* A precision score of about 0.76
* A recall of about 0.67

We can also evaluate each model individually

In [67]:
predictor.leaderboard(test)

model  score_test  score_val eval_metric  pred_time_test  \
0   WeightedEnsemble_L2    0.718351   0.725681          f1        2.726342   
1               XGBoost    0.716505   0.714426          f1        0.369850   
2              CatBoost    0.715429   0.701195          f1        0.076194   
3              LightGBM    0.711232   0.707510          f1        0.160882   
4         LightGBMLarge    0.706709   0.704390          f1        0.309829   
5            LightGBMXT    0.701381   0.693069          f1        0.305882   
6       NeuralNetFastAI    0.700163   0.692511          f1        0.293660   
7        NeuralNetTorch    0.684431   0.683206          f1        0.078300   
8      RandomForestGini    0.680092   0.689520          f1        0.568230   
9      RandomForestEntr    0.678169   0.683845          f1        0.547796   
10       ExtraTreesEntr    0.665286   0.666008          f1        0.884167   
11       ExtraTreesGini    0.662833   0.667327          f1        0.788824   
12       KNeighborsDist    0.441255   0.424051          f1        0.078404   
13       KNeighborsUnif    0.395336   0.378698          f1        0.079453   

    pred_time_val    fit_time  pred_time_test_marginal  \
0        0.983975  168.599248                 0.005664   
1        0.079966    9.025889                 0.369850   
2        0.030191   66.034918                 0.076194   
3        0.041574    2.766567                 0.160882   
4        0.091530    4.075467                 0.309829   
5        0.066486    3.741347                 0.305882   
6        0.063648   28.114774                 0.293660   
7        0.053046   41.236620                 0.078300   
8        0.177135    5.921120                 0.568230   
9        0.193663    7.545258                 0.547796   
10       0.423944    5.473397                 0.884167   
11       0.229027    4.239522                 0.788824   
12       0.033588    0.124770                 0.078404   
13       0.070191    0.132563                 0.079453   

    pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  \
0                 0.006414           4.326358            2       True   
1                 0.079966           9.025889            1       True   
2                 0.030191          66.034918            1       True   
3                 0.041574           2.766567            1       True   
4                 0.091530           4.075467            1       True   
5                 0.066486           3.741347            1       True   
6                 0.063648          28.114774            1       True   
7                 0.053046          41.236620            1       True   
8                 0.177135           5.921120            1       True   
9                 0.193663           7.545258            1       True   
10                0.423944           5.473397            1       True   
11                0.229027           4.239522            1       True   
12                0.033588           0.124770            1       True   
13                0.070191           0.132563            1       True   

    fit_order  
0          14  
1          11  
2           7  
3           4  
4          13  
5           3  
6          10  
7          12  
8           5  
9           6  
10          9  
11          8  
12          2  
13          1

References

[Salary prediction dataset](https://www.kaggle.com/datasets/ayessa/salary-prediction-classification)

https://auto.gluon.ai/stable/tutorials/tabular/tabular-essentials.html#presets):